stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -2786,6 +2786,25 @@ extension Dictionary : Equatable where Value : Equatable {
   }
 }
 
+extension Dictionary : Hashable where Value : Hashable {
+  /// The hash value for the dictionary.
+  ///
+  /// Two dictionaries that are equal will always have equal hash values.
+  ///
+  /// Hash values are not guaranteed to be equal across different executions of
+  /// your program. Do not save hash values to use during a future execution.
+  @_inlineable // FIXME(sil-serialize-all)
+  public var hashValue: Int {
+    // FIXME(ABI)#177: <rdar://problem/18915294> Issue applies to Dictionary too
+    var result: Int = _mixInt(0)
+    for (k, v) in self {
+      let combined = _combineHashValues(k.hashValue, v.hashValue)
+      result ^= _mixInt(combined)
+    }
+    return result
```

Comment:
Hi @lorentey ,
In the current version I removed _mixInt from Array types and from Optional as per your advise.
I am not 100% certain about your recommendation for the Dictionary - should I do the initial _mixInt(0) and the _mixInt(combined) - or are both of these superfluous too (referring to your discussion with @xwu )?
Sincerely,
/morten

Reply:
The `_mixInt(0)` is safe to eliminate, but we definitely need the `_mixInt(combined)`. 

You can think of `_mixInt` as a pseudo-random number generator that gives you a different 64-bit random number for every integer on its input. Calling it before the xor operation ensures that the xor won't cancel out bits in case all the elements return hash values within the same limited range. For example, it turns 

```
0 ^ 1 ^ 2 ^ 3 == 0
```

into

```
_mixInt(0) ^ _mixInt(1) ^ _mixInt(2) ^ _mixInt(3)
== 0xb2b24f688dc4164d ^ 0x792e33eb068557de ^ 0x66442b5dc51f5134 ^ 0xfbcb7d4146e1660d
== 0x56132a9f08bf76aa,
```

which looks like a much healthier hash value. (`Dictionary` then does another round of `_mixInt`, turning it into `0x16bb592381f12d84`, which isn't necessary in this case, but it doesn't really hurt, either.)